spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/code/InMemoryAuthorizationCodeServices.java

```diff
@@ -11,18 +12,52 @@
  * @author Dave Syer
  */
 public class InMemoryAuthorizationCodeServices extends RandomValueAuthorizationCodeServices {
+	/**
+	 *  From RFC6749: A maximum authorization code lifetime of 10 minutes is RECOMMENDED
+	 *
+	 *  @see <a href="https://tools.ietf.org/html/rfc6749#section-4.1.2">RFC6749 4.1.2.  Authorization Response</a>
+	 */
+	private static final int DEFAULT_CODE_LIFETIME_SECONDS = 10*60;
 
-	protected final ConcurrentHashMap<String, OAuth2Authentication> authorizationCodeStore = new ConcurrentHashMap<String, OAuth2Authentication>();
+	private int codeLiftetimeSeconds = DEFAULT_CODE_LIFETIME_SECONDS;
+
+	protected final ConcurrentHashMap<String, ExpiringOAuth2Authentication> authorizationCodeStore = new ConcurrentHashMap<String, ExpiringOAuth2Authentication>();
 
 	@Override
 	protected void store(String code, OAuth2Authentication authentication) {
-		this.authorizationCodeStore.put(code, authentication);
+		this.authorizationCodeStore.put(code, new ExpiringOAuth2Authentication(System.currentTimeMillis(),
+			authentication));
 	}
 
 	@Override
 	public OAuth2Authentication remove(String code) {
-		OAuth2Authentication auth = this.authorizationCodeStore.remove(code);
-		return auth;
+		removeExpired();
+		ExpiringOAuth2Authentication expiringOAuth2Authentication = this.authorizationCodeStore.remove(code);
+		return expiringOAuth2Authentication == null ? null : expiringOAuth2Authentication.authentication;
+	}
+
+	private void removeExpired() {
+		synchronized (authorizationCodeStore) {
+			long expirationThreshold = System.currentTimeMillis() - codeLiftetimeSeconds;
+			for (Iterator<ExpiringOAuth2Authentication> iterator = authorizationCodeStore.values().iterator(); iterator.hasNext(); ) {
+				if (iterator.next().createdTimestamp < expirationThreshold) {
+					iterator.remove();
+				}
+			}
+		}
 	}
 
+	private static class ExpiringOAuth2Authentication {
+		private final long createdTimestamp;
+		private final OAuth2Authentication authentication;
+
+		private ExpiringOAuth2Authentication(final long createdTimestamp, final OAuth2Authentication authentication) {
+			this.createdTimestamp = createdTimestamp;
+			this.authentication = authentication;
+		}
+	}
+
+	public void setCodeLiftetimeSeconds(int codeLiftetimeSeconds) {
```

Comment:
If possible I prefer following signature `(long lifetime, TimeUnit lifetimeTimeUnit)` (or whatever for naming) but is less errors prone. Some example from other library (*Apache HttpClient*) https://github.com/apache/httpclient/blob/a287ab63878f5e5a47f9cde639f04bb9e370478e/httpclient5/src/main/java/org/apache/hc/client5/http/impl/sync/HttpClientBuilder.java#L803